### PR TITLE
Use libmodbus to decode incoming Modbus TCP messages

### LINKED ISSUE #253: Modbus should accept multiple ADUs per TCP frame
I have a Modbus client that is sending multiple commands with the click of a button to the Modbus server of OpenPLC. Sometimes, some of the last Modbus commands are not executed by OpenPLC.

### Details
According to the [Modbus TCP specification](https://www.modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf), a Modbus message is the combination of exactly one ADU per TCP frame, which contains exactly one PDU.
<details><summary>Screenshots from the spec</summary>

Modbus TCP message diagram
Modbus implementation rules, specifying that "A TCP frame must transport only one MODBUS ADU. It is advised against sending multiple MODBUS requests or responses on the same TCP PDU"
</details> 

My client is a simple nodejs program that runs on a Windows machine and uses [`modbus-serial`](https://www.npmjs.com/package/modbus-serial) to communicate with the OpenPLC Modbus server. And unfortunately, my client is sometimes sending multiple ADUs per TCP frame.

<details><summary>Screenshot of Wireshark capture</summary>

A captured TCP frame with two ADUs
</details>

### Why this should be fixed in OpenPLC and not on the client
Now I know the OpenPLC Modbus server is conforming to the Modbus specification and it's the client that is not conformant (there is even [an open but stagnant issue](https://github.com/yaacov/node-modbus-serial/issues/540)), but there are multiple reasons why I think this should be handled on the OpenPLC side:

1. The spec is not clear. On the one side, it specifically forbids multiple ADUs in one TCP frame:
    > A TCP frame must transport only one MODBUS ADU.
   
   But then in the next sentence it seems to allow the possibility of sending multiple "Modbus requests" in one "TCP PDU":
   > It is advised against sending multiple MODBUS requests or responses on the same TCP PDU.

   I'm pretty sure a "TCP PDU" is a TCP frame. I initially thought a "MODBUS requests" request refers to a Modbus PDU, but the possibility and framework of having multiple PDUs per ADU is mentioned nowhere else in the spec, so I believe a "MODBUS requests" should be a ADU, which seems to directly contradict the first sentence.
2. The requirement to have only one ADU per TCP frame is incredible hard to implement on a system with an operating system that has a TCP stack. TCP is first and foremost a *streaming* protocol with no real notion of packet boundaries, framing is an implementation detail that is not exposed to the layer above. None of the operating systems provide enough control over the framing process to guarantee not merging frames, so the only option is to bypass the TCP stack completely and use some form of raw socket, but this means the client has to implement most parts of the TCP protocol themselves, which seems unreasonable.
3. The OpenPLC Modbus server is actually also not spec compliant because it has exactly the same issue when responding to requests.  Since the server is just using a regular socket, packets can be merged by the underlying operating system. Also, while [the `read`](https://github.com/thiagoralves/OpenPLC_v3/blob/3c08a69a6577859e14833d4b3df5974a0f356585/webserver/core/server.cpp#L174) seems to return the data one TCP frame at a time, there is actually no guarantee about that.
4. I think OpenPLC should follow the robustness principle "be conservative in what you send, be liberal in what you accept". Now we can't easily fix the sending part, but we can be more lenient in what we accept.
5. In addition to the nodejs library I tested two other Modbus client implementations ([`uModbus` (Python)](https://github.com/AdvancedClimateSystems/umodbus) and [`libmodbus` (C)](https://libmodbus.org)) that are directly linked [from the official Modbus website](https://www.modbus.org/tech.php) and I found that they also suffer from the same problem and can send multiple ADUs per TCP frame. I didn't invest the time to check the other implementations linked on that page, but I suspect they are not conformant too. So any client using these implementations would have this problem with OpenPLC.

### Proposal
The Modbus implementation of OpenPLC should treat the incoming data as a stream and ignore TCP frame boundaries. We can use a ringbuffer to archive this. When more data becomes available, the server should try and execute multiple ADUs from the buffer until there is not enough data in the buffer to parse the next Modbus ADU.

I am willing to work on a PR if you agree with this path forward.

### PR DESCRIPTION
**Pull request by Abestanis**
This allows us to better handle multiple Modbus ADUs in a single TCP frame (fixes #253).

`libmodbus` is better at parsing ADUs and we already use it when we are the Modbus client.

I tested this by sending many commands in a short amount of time and verifying via Wireshark that some of them were sent in one TCP frame. I made sure that OpenPLC was sending one response per request, even for those in one TCP packet.

In addition to the receiving problem, I also noticed that the result of `write` was ignored, which returns the number of bytes written. If we generate a lot of response messages, we could fill up the kernel buffer and drop some bytes. The new approach calls write in a loop until all bytes have been written or write returns an error.

### CODE REVIEW

**Comment by thiagoralves** [comment]
Hi @Abestanis 
Sorry for taking so long to look into this. I appreciate your PR, but I would rather avoid changing the entire Modbus implementation at this point. The included libmodbus on OpenPLC is quite old, and I plan to either update that or ditch libmodbus entirely. I have had issues with libmodbus in the past, and feel it is a bit too big (cumbersome might be a better word) for what we actually need in OpenPLC. I guess for the fix you need, simply splitting the received packet into separate Modbus PDUs and processing them individually might do the trick. You can use the frame size field to calculate the message size and split it there.

**Comment by LeSpocky** [comment]
> I guess for the fix you need, simply splitting the received packet into separate Modbus PDUs and processing them individually might do the trick. You can use the frame size field to calculate the message size and split it there.

This is usually not enough when handling fragmentation in a TCP stream.  Packets can be split anywhere, not necessarily at higher level protocol frame borders.  In other words: TCP fragmentation might happen inside of a Modbus PDU, so you would get first part of the PDU in one read from the TCP socket, and the second half of the PDU on the next read from the TCP socket.  A robust implementation must consider this.  (Been there, done that, for other protocols though.)

**Comment by Abestanis** [comment]
Hey @thiagoralves, no worries and sorry for the slow response, I've been quite busy the last few months.

>  I have had issues with libmodbus in the past, and feel it is a bit too big ...

I get what you are saying but consider this: It might seem like a big black box of code that you're not sure to trust, but it is much more likely to handle edge cases like #253, because it has been used by many people already. Implementing everything yourself is an option but you are also more likely to get things wrong and not notice it. Additionally, I think it can be problematic trying to re-implement everything. The main goal of this project is to provide an awesome PLC platform, and Modbus is just one of the underlying building blocks whose implementation should be left to those who want to focus and spend resources on providing a good Modbus implementation, which in this case is libmodbus.

> I guess for the fix you need, simply splitting the received packet into separate Modbus PDUs and processing them individually might do the trick.

I completely agree with LeSpocky about this, it won't be enough. TCP is a streaming protocol and makes no guarantees about packet boundaries at all. The behavior of not splitting Modbus packets between TCP frames we're seeing right now is an implementation detail of the TCP stack of my operating system, which could change at any time and might be different on another operating system.

Now we could change our modbus parser to only read enough data for the modbus header first, parse that, then only read the number of bytes as indicated in the header, but don't fail on incomplete packages and instead return the number of bytes that need to be read for the complete packet instead, and then loop until we received enough data like libmodbus is doing [here](https://github.com/stephane/libmodbus/blob/bb1344325025e59bb33144394bdacd8298283834/src/modbus.c#L414-L521). But this would be a lot of work and also basically means we would be changing the entire Modbus implementation, which we are both trying to avoid. With the current approach we have minimal changes to our code but still benefit from the battle proven code from libmodbus.

So TLDR, I would ask you to please reconsider your opinion on using libmodbus to solve this problem. Implementing the correct packet parsing manually is going to require too much work and I am unfortunately unable to spend the time that would be required for this. I've been running a controller with this patch for the last three months now and it has completely fixed the issue described in #253.

**Comment by thiagoralves** [comment]
Okay, I'm convinced! :) I'll merge the changes, and if things go south we can always revert it back. Thanks for working on this @Abestanis

**Comment by thiagoralves** [comment]
However @Abestanis correct me if I'm wrong, I just skimmed through your code changes, and it doesn't seem to be connecting the libmodbus parsing with the OpenPLC buffers. You're still calling the processModbus function, just avoiding to write the response of it back to the requester. Then you're creating a modbus socket using libmodbus and just providing a plain response. How is that tied with the actual state of internal registers and coils?

**Comment by Abestanis** [comment]
> Okay, I'm convinced! :) I'll merge the changes, and if things go south we can always revert it back.

Awesome, thanks you! 🎉

> You're still calling the processModbus function, just avoiding to write the response of it back to the requester

The changes in `processMessage` are not related to the main problem of #253, sorry for the confusion, maybe I should have split it into two PRs. They solve this potential problem:

> In addition to the receiving problem, I also noticed that the result of write was ignored, which returns the number of bytes written. If we generate a lot of response messages, we could fill up the kernel buffer and drop some bytes. The new approach calls write in a loop until all bytes have been written or write returns an error.

The actual fix to #253 is as follows:

Instead of just calling `listenToClient`, which just calls `read` on the `client_fd` and expects to read one whole message into the `buffer` but actually just reads up to `NET_BUFFER_SIZE`, we now:
1. Construct and initialize a modbus context, which contains the parsing machinery of libmodbus:
    https://github.com/thiagoralves/OpenPLC_v3/blob/5bf7feaddcd3959e528ff514645b799b2cc2f1cf/webserver/core/server.cpp#L226
2. We configure the context to read from our `client_fd` socket.
    https://github.com/thiagoralves/OpenPLC_v3/blob/5bf7feaddcd3959e528ff514645b799b2cc2f1cf/webserver/core/server.cpp#L227
3. Instead of `listenToClient` we then call `modbus_receive`, which reads from our `client_fd` until it has read *one full* modbus message into the `buffer`.
    https://github.com/thiagoralves/OpenPLC_v3/blob/5bf7feaddcd3959e528ff514645b799b2cc2f1cf/webserver/core/server.cpp#L239

`modbus_receive` eventually ends up [here](https://github.com/stephane/libmodbus/blob/bb1344325025e59bb33144394bdacd8298283834/src/modbus.c#L414-L521) which is the main loop that handles reading the modbus packet.

> How is that tied with the actual state of internal registers and coils?

It's not, that part is handled in `processMessage` and `processModbusMessage` the same way as before. The only difference is reading the message into `buffer`.

**Comment by Abestanis** [comment]
Please let me know if you want me to split the writing part into a separate PR.

**Comment by thiagoralves** [comment]
All merged now! :)

**Comment by thiagoralves** [comment]
I had to revert back this PR as it broke the OpenPLC debugger.

**Comment by Abestanis** [comment]
:( Do you know how this broke the debugger? I'm not familiar with the debugger functionality, does the debugger send modbus requests or how would it be affected by this?

**Comment by thiagoralves** [comment]
The debugger uses custom Modbus functions codes to exchange data with the runtime. I didn’t have time to dive into the code to understand what exactly broke it, but I suspect that it could be related to libmodbus not being able to handle the custom codes.

**Comment by Abestanis** [comment]
> but I suspect that it could be related to libmodbus not being able to handle the custom codes.

Indeed, ok, looks like we can't use libmodbus after all, which is a shame. Ok, then I'll try to take their approach and make it work with our implementation in a new MR.
